Default Paragraph Behavior

tadean opened the pull request:
This PR adds a default paragraph behavior plugin to serve as the base for future modification.

## Description
This pull request adds a default paragraph behavior plugin (**AZDefaultParagraphBehavior**) to serve as a base class for paragraph styles. This is intended to contain the global paragraph styles that are potentially available to all paragraphs.
Currently it handles adding a default bottom spacing element to paragraphs.

Previous functionality that handled loading paragraph-specific css and js libraries has been refactored and placed into
this plugin as well.

The new behavior is enabled on the existing **az_text** paragraph.

The existing paragraph test is altered to check that the plugin is working.

Contributing documentation is updated.

## Related Issue
#331 

## How Has This Been Tested?
Create a text paragraph. Verify that the Additional Options detail pane exists and contains the Bottom Spacing dropdown.Verify that the selected spacing element is added to the paragraph in rendering.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added this Pull Request to the AZ Quickstart project in the right column.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tadean commented:
![image](https://user-images.githubusercontent.com/51835841/94715288-1dd13580-0302-11eb-8d43-74a6a87f1272.png)

![image](https://user-images.githubusercontent.com/51835841/94715314-26297080-0302-11eb-89c4-df864db87f19.png)

mmunro-ltrr reviewed (approved):
After messing up the live demo, it's only polite to approve this!

joeparsons commented on modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
I think the preferred way of injecting additional dependencies (beyond what the parent class does is to actually override the `__construct()` method with additional typeHinted dependencies in the method signature, and then inject all of the dependencies (both those defined by the parent and the additional ones defined in the extended class) in the `create()` method.  We have an example of this in the az_quickstart_core module here:
https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_core/src/Form/QuickstartCoreSettingsForm.php#L31-L57

joeparsons replied:
@tadean This documentation on drupal.org provides some similar examples: https://www.drupal.org/docs/8/api/services-and-dependency-injection/dependency-injection-for-a-form

tadean replied:
Thanks @joeparsons ! Updated to the style of overriding that overrides the constructor.

camikazegreen commented:
In the help text for the Bottom Spacing, can we link directly to: https://digital.arizona.edu/arizona-bootstrap/docs/2.0/utilities/spacing

joeparsons commented:
This looks AMAZING overall!

joeparsons commented on modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php:
```diff
@@ -0,0 +1,154 @@
+ *   weight = 0
+ * )
+ */
+class AZDefaultParagraphBehavior extends ParagraphsBehaviorBase {
```
Nitpick:  Would it be better to name our behavior `AZDefaultParagraphsBehavior` (with an `s`).  It seems that within the Paragraphs module, behaviors are always referred to as Paragraphs Behaviors.

tadean replied:
I agree, that seems better. Updated to change the name to `AZDefaultParagraphsBehavior`.

trackleft reviewed (approved):
Magic!

tadean commented:
Updated with the following:

- Remove `az_` scoping on FAPI elements inside the `az_display_settings` element.
- Update style of container dependency injection.
- Remove a bare call to `t()`
- Added link to spacing documentation
- Changed name of class to `AZDefaultParagraphsBehavior`
